homeassistant/components/evohome/__init__.py

```diff
@@ -118,91 +113,159 @@
+            await self._store.async_save({})
+            return
+
+        # evohomeasync2 uses naive/local datetimes
+        access_token_expires = dt_local_to_aware(
+            self.client_v2.access_token_expires  # type: ignore[arg-type]
+        )
+
+        self._tokens = {
+            CONF_USERNAME: self.client_v2.username,
+            REFRESH_TOKEN: self.client_v2.refresh_token,
+            ACCESS_TOKEN: self.client_v2.access_token,
+            ACCESS_TOKEN_EXPIRES: access_token_expires.isoformat(),
+        }
+
+        self.session_id = self.client_v1.broker.session_id if self.client_v1 else None
+
+        app_storage = self._tokens
+        if self.client_v1:
+            app_storage[USER_DATA] = {SZ_SESSION_ID: self.session_id}
+
+        await self._store.async_save(app_storage)
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up the Evohome integration."""
+
+    sess = EvoSession(hass)
 
     try:
-        await client_v2.login()
+        await sess.authenticate(
+            config[DOMAIN][CONF_USERNAME],
+            config[DOMAIN][CONF_PASSWORD],
+        )
+
     except evo.AuthenticationFailed as err:
         handle_evo_exception(err)
         return False
+
     finally:
         config[DOMAIN][CONF_PASSWORD] = "REDACTED"
 
-    assert isinstance(client_v2.installation_info, list)  # mypy
+    broker = EvoBroker(sess)
 
-    loc_idx = config[DOMAIN][CONF_LOCATION_IDX]
-    try:
-        loc_config = client_v2.installation_info[loc_idx]
-    except IndexError:
-        _LOGGER.error(
-            (
-                "Config error: '%s' = %s, but the valid range is 0-%s. "
-                "Unable to continue. Fix any configuration errors and restart HA"
-            ),
-            CONF_LOCATION_IDX,
-            loc_idx,
-            len(client_v2.installation_info) - 1,
-        )
+    if not broker.validate_location(
+        config[DOMAIN][CONF_LOCATION_IDX],
+    ):
         return False
 
-    if _LOGGER.isEnabledFor(logging.DEBUG):
-        loc_info = {
-            SZ_LOCATION_ID: loc_config[SZ_LOCATION_INFO][SZ_LOCATION_ID],
-            SZ_TIME_ZONE: loc_config[SZ_LOCATION_INFO][SZ_TIME_ZONE],
-        }
-        gwy_info = {
-            SZ_GATEWAY_ID: loc_config[GWS][0][SZ_GATEWAY_INFO][SZ_GATEWAY_ID],
-            TCS: loc_config[GWS][0][TCS],
-        }
-        _config = {
-            SZ_LOCATION_INFO: loc_info,
-            GWS: [{SZ_GATEWAY_INFO: gwy_info}],
-        }
-        _LOGGER.debug("Config = %s", _config)
-
-    client_v1 = ev1.EvohomeClient(
-        client_v2.username,
-        client_v2.password,
-        session_id=user_data.get(SZ_SESSION_ID) if user_data else None,  # STORAGE_VER 1
-        session=async_get_clientsession(hass),
-    )
-
     hass.data[DOMAIN] = {}
-    hass.data[DOMAIN]["broker"] = broker = EvoBroker(
-        hass, client_v2, client_v1, store, config[DOMAIN]
+    hass.data[DOMAIN]["broker"] = broker
```

Comment:
how about using set default here

Reply:
But this integration will never be **reloaded** - or have I missed something?

Reply:
Just to clean up the number of steps to populate the dictionary.

Reply:
```python
hass.data[DOMAIN] = {"broker": broker}
```

Reply:
How about switching to runtime_data instead.

Reply:
That is the long-term goal. I plan to do so when I implement config flow.